### PR TITLE
Migrate codegen to operate on orc::ThreadSafeModule

### PR DESCRIPTION
**Pull request by pchintalapudi**
orc::ThreadSafeModule and orc::ThreadSafeContext use reference-counted contexts to allow multiple contexts to be created and destroyed safely during codegen. Moving TSModule to be our API boundary gets us closer to parallelizing codegen without leaking contexts.

Depends on ~~#43770 to remove the global debug information with associated global context~~ #44454 to refactor a few globals.

### CODE REVIEW

**Comment by pchintalapudi** [comment]
There's a few indentation TODOs left in this PR, which I've deliberately left out for now as they cause large whitespace diffs that confuse the diff viewer.

**Comment by pchintalapudi** [comment]
Per an offline discussion with Valentin, we will change the interface of `jl_create_native` to accept a TSModule to codegen into rather than creating its own TSModule with a provided TSContext. This allows users of `jl_create_native` to specify their own data layout and targets for the destination module.

File: src/aotcompile.cpp [hunk 1]
```diff
@@ -256,19 +246,24 @@ static void jl_ci_cache_lookup(const jl_cgparams_t &cgparams, jl_method_instance
...
         cgparams = &jl_default_cgparams;
     jl_native_code_desc_t *data = new jl_native_code_desc_t;
-    jl_codegen_params_t params;
-    params.params = cgparams;
-    std::map<jl_code_instance_t*, jl_compile_result_t> emitted;
+    orc::ThreadSafeModule backing;
+    if (!llvmmod) {
+        backing = jl_create_llvm_module("text", jl_ExecutionEngine->getContext(), cgparams ? cgparams : &jl_default_cgparams);
+    }
+    orc::ThreadSafeModule &clone = llvmmod ? *reinterpret_cast<orc::ThreadSafeModule*>(llvmmod) : backing;
```
Comment (vchuravy) [inline src/aotcompile.cpp:258]: Same should be an unwrap?

File: src/aotcompile.cpp [hunk 2]
```diff
@@ -508,8 +511,8 @@ void jl_dump_native_impl(void *native_code,
             jl_safe_printf("ERROR: target does not support generation of object files\n");
 
     // Reset the target triple to make sure it matches the new target machine
-    data->M->setTargetTriple(TM->getTargetTriple().str());
-    data->M->setDataLayout(create_jl_data_layout(*TM));
+    data->M.getModuleUnlocked()->setTargetTriple(TM->getTargetTriple().str());
+    data->M.getModuleUnlocked()->setDataLayout(create_jl_data_layout(*TM));
```
Comment (vchuravy) [inline src/aotcompile.cpp:515]: Can we export `create_jl_datalayout` as `jl_create_datalayout`?

File: src/aotcompile.cpp [hunk 3]
```diff
@@ -519,13 +522,13 @@ void jl_dump_native_impl(void *native_code,
 
     // add metadata information
     if (imaging_mode) {
-        emit_offset_table(*data->M, data->jl_sysimg_gvars, "jl_sysimg_gvars", T_psize);
-        emit_offset_table(*data->M, data->jl_sysimg_fvars, "jl_sysimg_fvars", T_psize);
+        emit_offset_table(*data->M.getModuleUnlocked(), data->jl_sysimg_gvars, "jl_sysimg_gvars", T_psize);
```
Comment (vchuravy) [inline src/aotcompile.cpp:525]: Would it make sense to get M once and then pass it in?

**Review by vtjnash — APPROVED** [review]

**Comment by aviatesk** [comment]
@gbaraldi bisected that this commit introduced the error reported at https://github.com/JuliaLang/julia/issues/45302.

**Comment by pchintalapudi** [comment]
That's interesting, I wouldn't be surprised that there's a hidden bug given the size of the changeset but the changes to jl_merge_module aren't actually meaningful.

**Comment by pchintalapudi** [comment]
Looks like at least the error in jl_merge_module during LoopVectorization tests is:
```
Attempting to replace ; Function Attrs: alwaysinline
declare linkonce_odr i32 @"julia_readraw_turbo!_14765u14814"(i32, i32) #4

with global function  ; Function Attrs: alwaysinline
define linkonce_odr i32 @"julia_readraw_turbo!_14765u14814"(i32 %0, i32 %1, i32 %2) #0 {
top:
  %res = call i32 @llvm.x86.bmi.bzhi.32(i32 %0, i32 %1)
  ret i32 %res
}
```

**Comment by pchintalapudi** [comment]
And another one:
```
Attempting to replace ; Function Attrs: alwaysinline
declare linkonce_odr i32 @"julia_rshift_i_avx!_47472u47499"(i32, i32) #3

with global function  ; Function Attrs: alwaysinline
define linkonce_odr i32 @"julia_rshift_i_avx!_47472u47499"(i32 %0, i32 %1, i32 %2) #0 {
top:
  %res = call i32 @llvm.x86.bmi.bzhi.32(i32 %0, i32 %1)
  ret i32 %res
}
```

Looks like there's a common x86.bmi.bzhi intrinsic that's causing some issues between function definition/declaration

**Comment by pchintalapudi** [comment]
Reduced to
```
using LoopVectorization
img1 = Matrix{UInt8}(undef, 10, 10)
raw = rand(UInt8, (30 * 10) ÷ 4)
function readraw_turbo!(img, raw)
         npack = length(raw) ÷ 3
         @turbo for i = 0:npack-1
           img[1+4i] = raw[2+3i] << 4
           img[2+4i] = raw[1+3i]
           img[3+4i] = raw[2+3i]
           img[4+4i] = raw[3+3i]
         end
         img
       end
readraw_turbo!(img1,raw)
```

**Comment by pchintalapudi** [comment]
And again to
```
using LoopVectorization;img1 = Matrix{UInt8}(undef, 10, 10);raw = rand(UInt8, (30 * 10) ÷ 4);function readraw_turbo!(img, raw)
         npack = length(raw) ÷ 3
         @turbo for i = 0:npack-1
           img[1+4i] = raw[2+3i] << 4
         end
         img
       end;readraw_turbo!(img1,raw)
```

**Comment by pchintalapudi** [comment]
I am now slightly suspicious of this line: https://github.com/JuliaSIMD/VectorizationBase.jl/blob/9ae47c5c12c7bec9b57c132c24f67c4f14090c9e/src/llvm_intrin/masks.jl#L336

`llvmcall_expr(decl, instrs, T, :(Tuple{$T,$T}), typ, [typ, typ, typ], [:a, :b])`

I'm not sure, but I'm guessing that the list should only have 2 typ, not 3

**Pull request merged**